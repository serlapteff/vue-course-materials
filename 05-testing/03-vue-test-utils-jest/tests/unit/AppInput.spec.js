const { mount } = require('@vue/test-utils');
// or import { shallowMount } from '@vue/test-utils';
import AppInput from '@/components/AppInput';

describe('AppInput', () => {
  it('should be defined', () => {
    expect(AppInput).toBeDefined();
  });

  describe('without initial props', () => {
    let wrapper;
    let inputGroup;
    let formControl;

    beforeEach(() => {
      wrapper = mount(AppInput);
      inputGroup = wrapper.find('.input-group');
      formControl = inputGroup.find('.form-control');
    });

    it('should render input.input-group inside form-control', async () => {
      expect(inputGroup.exists()).toBe(true);
      expect(formControl.exists()).toBe(true);
    });

    it('should have form-control_rounded on input only when rounded is true', async () => {
      expect(formControl.classes('form-control_rounded')).toBe(false);
      await wrapper.setProps({ rounded: true });
      // await wrapper.vm.$nextTick();
      expect(formControl.classes('form-control_rounded')).toBe(true);
    });

    it('should have current value', async () => {
      const value = 'SameValue';
      await wrapper.setProps({ value });
      expect(formControl.element.value).toBe(value);
    });

    it('should match snapshot', async () => {
      expect(wrapper).toMatchSnapshot();
    });
  });

  it.each(['password', 'num', 'range', 'date', 'time', 'email', 'tel'])(
    'should render current input type (%s)',
    (type) => {
      const wrapper = mount(AppInput, {
        propsData: { type },
      });
      expect(wrapper.find('.form-control').attributes('type')).toBe(type);
    },
  );

  it('should handle native input event and emit input with inputted value', async () => {
    const value = 'SomeText';
    const handler = jest.fn();
    const wrapper = mount(AppInput, {
      listeners: {
        input: (event) => handler(event),
      },
    });
    // wrapper.find('.form-control').element.value = value;
    // await wrapper.find('.form-control').trigger('input');
    await wrapper.find('.form-control').setValue(value);

    expect(handler).toHaveBeenCalled();
    expect(handler).toHaveBeenCalledWith(value);
    // or
    expect(wrapper.emitted().input).toBeTruthy();
    expect(wrapper.emitted().input).toHaveLength(1);
    expect(wrapper.emitted().input[0]).toEqual([value]);
  });
});
