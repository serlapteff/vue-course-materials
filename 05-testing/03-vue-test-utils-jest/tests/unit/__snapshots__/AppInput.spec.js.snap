// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppInput without initial props should match snapshot 1`] = `<div class="input-group"><input class="form-control"></div>`;
